asmcomp/cmm_helpers.ml

```diff
@@ -217,16 +216,49 @@ let lsr_int c1 c2 dbg =
 
 let asr_int c1 c2 dbg =
   match c2 with
-    Cconst_int (0, _) ->
+  | Cconst_int (0, _) ->
       c1
   | Cconst_int (n, _) when n > 0 ->
       Cop(Casr, [ignore_low_bit_int c1; c2], dbg)
   | _ ->
       Cop(Casr, [c1; c2], dbg)
 
+let add_float c1 c2 dbg =
+  Cop(Caddf, [c1; c2], dbg)
+
+let sub_float c1 c2 dbg =
+  Cop(Csubf, [c1; c2], dbg)
+
+let mul_float c1 c2 dbg =
+  match (c1, c2) with
+  | (c, Cconst_float(2.0, _)) | (Cconst_float(2.0, _), c) ->
+     (* We transform x *. 2.0 into x +. x, by creating a temporary variable and
+        outputting let tmp = x in tmp +. tmp to avoid calculating x twice. *)
+     let tmp_var = V.create_local "tmp" in
+     let var = Cvar tmp_var in
+     Clet (VP.create tmp_var, c, add_float var var dbg)
+  | _, _ ->
+      Cop(Cmulf, [c1; c2], dbg)
+
+let div_float c1 c2 dbg =
+  match (c1, c2) with
+  | c, Cconst_float(f, _) ->
+      (* x = 0.5 if and only if f is a power of 2 *)
+      let x, exp = Float.frexp f in
+      (* We can transform x/.2^{N} into x*.2^{-N} if and only if |N| < 1023,
+         because otherwise one of the may not be a valid floating point number
+         (2.**1023. evaluates to +infinity). *)
+      if x = 0.5 && abs exp < 1023
```

Comment:
Is the bound correct? `f = x *. 2.0 ** exp`, so in your formula `N` is `exp - 1`, not `exp`.